Sources/Testing/Parameterization/CustomTestArgument.swift

```diff
@@ -0,0 +1,151 @@
+  ///   - context: The context in which the argument was passed.
+  ///
+  /// - Returns: `nil` if an ID cannot be formed from the specified test
+  ///   argument value.
+  ///
+  /// - Throws: Any error encountered while attempting to encode `value`.
+  ///
+  /// This function is not part of the public interface of the testing library.
+  ///
+  /// ## See Also
+  ///
+  /// - ``CustomTestArgumentEncodable``
+  init?(identifying value: some Sendable, in context: Test.Case.Argument.Context) throws {
+#if canImport(Foundation)
+    guard Configuration.current?.isTestArgumentEncodingEnabled ?? false else {
+      return nil
+    }
+
+    let encodableValue: (any Encodable)? = if let customEncodable = value as? any CustomTestArgumentEncodable {
+      _customArgumentWrapper(for: customEncodable, in: context)
+    } else if let encodable = value as? any Encodable {
+      encodable
+    } else if let identifiable = value as? any Identifiable, let encodableID = identifiable.id as? any Encodable {
+      encodableID
+    } else {
+      nil
+    }
+
+    guard let encodableValue else {
+      return nil
+    }
+
+    self = .init(bytes: try Self._encode(encodableValue))
+#else
+    nil
+#endif
+  }
+
+#if canImport(Foundation)
+  /// Encode the specified test argument value and store its encoded
+  /// representation as an array of bytes suitable for storing in an instance of
+  /// ``Test/Case/Argument/ID-swift.struct``.
+  ///
+  /// - Parameters:
+  ///   - value: The value to encode.
+  ///
+  /// - Returns: An array of bytes containing the encoded representation.
+  ///
+  /// - Throws: Any error encountered during encoding.
+  private static func _encode(_ value: some Encodable) throws -> [UInt8] {
+    let encoder = JSONEncoder()
+
+    // Keys must be sorted to ensure deterministic matching of encoded data.
+    encoder.outputFormatting.insert(.sortedKeys)
+
+    return .init(try encoder.encode(value))
+  }
+#endif
+}
+
+/// A encodable type which wraps a ``CustomTestArgumentEncodable`` value.
+private struct _CustomArgumentWrapper<T>: Encodable where T: CustomTestArgumentEncodable {
+  /// The value this instance wraps, which implements custom test argument
+  /// encoding logic.
+  var value: T
+
+  /// The context in which the custom test argument was used.
+  var context: Test.Case.Argument.Context
+
+  func encode(to encoder: any Encoder) throws {
+    try value.encodeTestArgument(to: encoder, in: context)
+  }
+}
+
+/// Create an encodable wrapper for a value which conforms to
+/// ``CustomTestArgumentEncodable``.
+///
+/// - Parameters:
+///   - value: The value which implements custom test argument encoding logic.
+///   - context: The context in which the custom test argument was used.
+///
+/// - Returns: An encodable wrapper for the specified value.
+private func _customArgumentWrapper(for value: some CustomTestArgumentEncodable, in context: Test.Case.Argument.Context) -> some Encodable {
```

Comment:
I think this function may be redundant? Can't you just call the constructor for `_CustomArgumentWrapper` directly?

Reply:
Tried that first but the compiler doesn't open the existential in an initializer, it seems

Reply:
You can make it a local function in the one place that calls it, I think. No need for a whole symbol to open an existential. (We do that in a few other spots.)